pallets/liquidity-pools-gateway/queue/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use std::fmt::Debug;
+
+use cfg_traits::liquidity_pools::{MessageProcessor, MessageQueue as MessageQueueT};
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use parity_scale_codec::FullCodec;
+use scale_info::TypeInfo;
+use sp_runtime::traits::{AtLeast32BitUnsigned, EnsureAdd, EnsureAddAssign, One};
+
+mod weights;
+
+use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The message type.
+		type Message: Clone + Debug + PartialEq + MaxEncodedLen + TypeInfo + FullCodec;
+
+		/// Type used for message identification.
+		type MessageNonce: Parameter
+			+ Member
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ EnsureAdd
+			+ MaybeSerializeDeserialize
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		/// Type used for processing messages.
+		type MessageProcessor: MessageProcessor<Message = Self::Message>;
+
+		/// Weight information.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn message_nonce_store)]
+	pub type MessageNonceStore<T: Config> = StorageValue<_, T::MessageNonce, ValueQuery>;
+
+	/// Storage for messages that will be processed during the `on_idle` hook.
+	#[pallet::storage]
+	#[pallet::getter(fn message_queue)]
+	pub type MessageQueue<T: Config> = StorageMap<_, Blake2_128Concat, T::MessageNonce, T::Message>;
+
+	/// Storage for messages that failed during processing.
+	#[pallet::storage]
+	#[pallet::getter(fn failed_message_queue)]
+	pub type FailedMessageQueue<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::MessageNonce, (T::Message, DispatchError)>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A message was submitted.
+		MessageSubmitted {
+			nonce: T::MessageNonce,
+			message: T::Message,
+		},
+
+		/// Message execution success.
+		MessageExecutionSuccess {
+			nonce: T::MessageNonce,
+			message: T::Message,
+		},
+
+		/// Message execution failure.
+		MessageExecutionFailure {
+			nonce: T::MessageNonce,
+			message: T::Message,
+			error: DispatchError,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Message not found.
+		MessageNotFound,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_now: BlockNumberFor<T>, max_weight: Weight) -> Weight {
+			Self::service_message_queue(max_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
```

Comment:
Curious, why a new pallet? Can not this behavior be embedded in the gateway?

Reply:
This was mentioned in one of the previous discussions around this topic, having the queue logic separately will make it easier for us to change it to `Tasks` in the future. Furthermore, it makes the gateway lighter and hopefully the only logic in there is the one dealing with messages and message proofs.

Reply:
I didn't read it. Thanks! makes sense 👍🏻